google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/QueryResult.java

```diff
@@ -16,71 +16,35 @@
 
 package com.google.cloud.bigquery;
 
-import com.google.cloud.PageImpl;
-import com.google.common.base.MoreObjects;
-import com.google.common.collect.ImmutableList;
+import static com.google.common.base.Preconditions.checkNotNull;
 
+import com.google.api.gax.paging.Page;
+import com.google.common.base.Function;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.Iterables;
 import java.util.Objects;
 
-public class QueryResult extends PageImpl<FieldValueList> {
+// TODO(pongad): Serializable ?
+public class QueryResult implements Page<FieldValueList> {
 
   private static final long serialVersionUID = -4831062717210349818L;
 
   private final Schema schema;
   private final long totalRows;
-
-  interface QueryResultsPageFetcher extends PageImpl.NextPageFetcher<FieldValueList> {
-
-    @Override
-    QueryResult getNextPage();
-  }
-
-  static final class Builder {
-
-    private QueryResultsPageFetcher pageFetcher;
-    private String cursor;
-    private Iterable<FieldValueList> results;
-    private Schema schema;
-    private long totalRows;
-
-    private Builder() {}
-
-    Builder setSchema(Schema schema) {
-      this.schema = schema;
-      return this;
-    }
-
-    Builder setTotalRows(long totalRows) {
-      this.totalRows = totalRows;
-      return this;
-    }
-
-    Builder setPageFetcher(QueryResultsPageFetcher pageFetcher) {
-      this.pageFetcher = pageFetcher;
-      return this;
-    }
-
-    Builder setCursor(String cursor) {
-      this.cursor = cursor;
-      return this;
-    }
-
-    Builder setResults(Iterable<FieldValueList> results) {
-      this.results = results;
-      return this;
-    }
-
-    QueryResult build() {
-      return new QueryResult(this);
-    }
-  }
-
-  private QueryResult(Builder builder) {
-    super(builder.pageFetcher,
-        builder.cursor,
-        builder.results != null ? builder.results : ImmutableList.<FieldValueList>of());
-    this.schema = builder.schema;
-    this.totalRows = builder.totalRows;
+  private final Page<FieldValueList> pageNoSchema;
+  private final Function<FieldValueList, FieldValueList> addSchemaFunc;
+
+  QueryResult(final Schema schema, long totalRows, Page<FieldValueList> pageNoSchema) {
```

Comment:
It might be good to get the `totalRows` from the call to `tabledata.list`, but this does introduce the issue that the total rows won't be populated until you start iterating.

Reply:
We should do this a little later. We want to also return `totalRows` from `listTableData` (method, not RPC) anyway. After we do that, it will be easy to do this.